# 1. LIBRARIES & DATA IMPORT

In [ ]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns
import os


In [29]:
from custom_eda_functions import *

In [4]:
# !TBD: Change the path with os library to create a relative file path (it is absolute atm)
df = pd.read_csv(r"C:\Users\ernes\Desktop\GITHUB\diabetes_prediction\assets\data\TAIPEI_diabetes.csv")

# Creation of a deep copy of the original df
df_deep_copy = df.copy()

---
# 2. INITIAL DATA OVERVIEW

#### This phase is dedicated to general understanding of the data available on hand and identification of the following:
* general shape of the dataframe;
* features available for the prediction of our target variable;
* detecting NULL data points;
* columns' data types;
* check if the data point respects the data type of the column;
* duplicate values.

In [5]:
df.shape

(15000, 10)

In [9]:
df.head()

PatientID  Pregnancies  PlasmaGlucose  DiastolicBloodPressure  \
0    1354778            0            171                      80   
1    1147438            8             92                      93   
2    1640031            7            115                      47   
3    1883350            9            103                      78   
4    1424119            1             85                      59   

   TricepsThickness  SerumInsulin        BMI  DiabetesPedigree  Age  Diabetic  
0                34            23  43.509726          1.213191   21         0  
1                47            36  21.240576          0.158365   23         0  
2                52            35  41.511523          0.079019   23         0  
3                25           304  29.582192          1.282870   43         1  
4                27            35  42.604536          0.549542   22         0

In [10]:
df.tail(10)

PatientID  Pregnancies  PlasmaGlucose  DiastolicBloodPressure  \
14990    1220763            5            169                      83   
14991    1603490            0            114                      53   
14992    1202654            3             48                      60   
14993    1165919            1            128                      59   
14994    1453189            0             72                      99   
14995    1490300           10             65                      60   
14996    1744410            2             73                      66   
14997    1742742            0             93                      89   
14998    1099353            0            132                      98   
14999    1386396            3            114                      65   

       TricepsThickness  SerumInsulin        BMI  DiabetesPedigree  Age  \
14990                31            60  49.004793          0.187397   53   
14991                34            40  44.760534          0.143487   23   
14992                24            81  29.417154          0.159605   42   
14993                21           182  19.766514          0.167280   53   
14994                32            32  20.932808          0.545038   22   
14995                46           177  33.512468          0.148327   41   
14996                27           168  30.132636          0.862252   38   
14997                43            57  18.690683          0.427049   24   
14998                18           161  19.791645          0.302257   23   
14999                47           512  36.215437          0.147363   34   

       Diabetic  
14990         1  
14991         0  
14992         1  
14993         0  
14994         0  
14995         1  
14996         1  
14997         0  
14998         0  
14999         1

In [ ]:
# First check of the NULL data points inside the columns and their data types
df.info()

In [ ]:
# Second check of the NULL values
df.isna().sum()

In [23]:
# Check if the data points data type corresponds to their columns' data type
def dtype_respect_per_column(df):
    for col in df.columns:
        print(f"{col}:")
        dtype_count = {}

        for i in df[col]:
            dtype_name = type(i).__name__
            if dtype_name in dtype_count:
                dtype_count[dtype_name] = dtype_count[dtype_name] + 1
            else:
                dtype_count[dtype_name] = 1

        for dtype_name in dtype_count:
            print(f"{dtype_name}: {str(dtype_count[dtype_name])}")
            print("\n")

dtype_respect_per_column(df)

PatientID:
int: 15000


Pregnancies:
int: 15000


PlasmaGlucose:
int: 15000


DiastolicBloodPressure:
int: 15000


TricepsThickness:
int: 15000


SerumInsulin:
int: 15000


BMI:
float: 15000


DiabetesPedigree:
float: 15000


Age:
int: 15000


Diabetic:
int: 15000




In [ ]:
# Check of the duplicated rows
duplicate_rows = df.duplicated().sum()

print(f"Total duplicated rows: {duplicate_rows}.")

Total duplicated rows: 0.


In [25]:
# Check of duplicated data points per column
def duplicate_cells_per_column(df):
    for col in df.columns:
        print(f"{col}: {df[col].duplicated().sum()}")

duplicate_cells_per_column(df)

PatientID: 105
Pregnancies: 14985
PlasmaGlucose: 14851
DiastolicBloodPressure: 14910
TricepsThickness: 14931
SerumInsulin: 14337
BMI: 0
DiabetesPedigree: 1
Age: 14944
Diabetic: 14998


In [ ]:
# Check of the top 10 most frequent values per column
for col in df:
    print(f"Top 10 most frequent values for {col}:")
    print(df[col].value_counts().head(10))
    print("\n")

Top 10 most frequent values for PatientID:
PatientID
1321191    2
1772038    2
1693955    2
1951340    2
1104393    2
1015338    2
1460338    2
1459465    2
1733432    2
1567493    2
Name: count, dtype: int64


Top 10 most frequent values for Pregnancies:
Pregnancies
0    4377
1    2864
3    1219
6    1070
2     939
7     895
9     881
5     687
4     675
8     645
Name: count, dtype: int64


Top 10 most frequent values for PlasmaGlucose:
PlasmaGlucose
97     263
96     240
118    209
95     199
107    199
119    196
93     195
85     194
117    193
114    191
Name: count, dtype: int64


Top 10 most frequent values for DiastolicBloodPressure:
DiastolicBloodPressure
81    445
79    416
78    406
83    404
80    401
86    400
84    394
82    385
60    358
85    354
Name: count, dtype: int64


Top 10 most frequent values for TricepsThickness:
TricepsThickness
11    657
9     593
10    572
45    521
34    517
7     511
8     493
44    488
31    481
35    481
Name: count, dtype: int64


Top

#### INITIAL DATA OVERVIEW OBSERVATIONS

The studied dataset consists of 15 000 records in total.

It has 10 columns, with all of them being of a numerical data type (8x int64 and 2x float64).

The features are of numeric nature, both continuous and discrete. The predictor is a categorical, binary data type.
All the data points match the expected data types for their respective columns. Based on the preliminary notes documentation, these data types seem correct, given the nature of the features. Therefore, there is no need for type casting in order to normalize the data types.

The dataset contains no NULL values or duplicate rows, allowing us to explore it further during the EDA phase, without additional manipulations at this stage.

An interesting observation is that 105 PatientID values are duplicated, suggesting that some patients may have undergone multiple check-ups.

2/3 of the dataset contains non-diabetic patients records (predictor variable), which may lead to a class imbalance, where a machine learning model is better at predicting the non-diabetic patients, rather than the diabetic ones. An approach that might be helpful in this case would be resampling.

---
# 3. EXPLORATORY DATA ANALYSIS (EDA)

#### This part is dedicated to answer the following:
* Formulate hypotheses based on observations and confirm or reject them.
* Explore the data more in detail using a statistical approach.
* Detect the outliers and decide on how to handle them.
* Understand if the data on hand is better at predicting the diabetic or non-diabetic outcome.
* Take a closer look at the repeated PatientID, to understand if some patient underwent multiple medial checkups.
* Explore the relationships between the features, and their correlation with the outcome predictor.
* Understand if we can and should develop additional features.
* Understand if encoding of the data is needed.

### 3.1 Basic statistics of the dataset

In [5]:
# Division of the dataset into diabetic and non-diabetic dataframes
df_non_diabetic = df[df["Diabetic"] == 0]
df_diabetic = df[df["Diabetic"] == 1]

In [7]:
# Descriptive statistics of the features of the full dataframe
df.describe().drop(columns=["PatientID"])

Pregnancies  PlasmaGlucose  DiastolicBloodPressure  TricepsThickness  \
count  15000.000000   15000.000000            15000.000000      15000.000000   
mean       3.224533     107.856867               71.220667         28.814000   
std        3.391020      31.981975               16.758716         14.555716   
min        0.000000      44.000000               24.000000          7.000000   
25%        0.000000      84.000000               58.000000         15.000000   
50%        2.000000     104.000000               72.000000         31.000000   
75%        6.000000     129.000000               85.000000         41.000000   
max       14.000000     192.000000              117.000000         93.000000   

       SerumInsulin           BMI  DiabetesPedigree           Age  \
count  15000.000000  15000.000000      15000.000000  15000.000000   
mean     137.852133     31.509646          0.398968     30.137733   
std      133.068252      9.759000          0.377944     12.089703   
min       14.000000     18.200512          0.078044     21.000000   
25%       39.000000     21.259887          0.137743     22.000000   
50%       83.000000     31.767940          0.200297     24.000000   
75%      195.000000     39.259692          0.616285     35.000000   
max      799.000000     56.034628          2.301594     77.000000   

           Diabetic  
count  15000.000000  
mean       0.333333  
std        0.471420  
min        0.000000  
25%        0.000000  
50%        0.000000  
75%        1.000000  
max        1.000000

In [50]:
# Number of records for diabetic vs non-diabetic patients
print(f"{" " * 34} Number of records for diabetic vs non-diabetic patients")
df.groupby("Diabetic").size().to_frame("Count")

                                   Number of records for diabetic vs non-diabetic patients


Count
Diabetic       
0         10000
1          5000

In [35]:
# Mean comparison of diabetic and non-diabetic patients
df.groupby("Diabetic").mean().drop(columns=["PatientID"]).style.set_caption("Mean comparison of diabetic and non-diabetic patients <br><br>")

In [34]:
# Standard deviation comparison of diabetic and non-diabetic patients
df.groupby("Diabetic").std().drop(columns=["PatientID"]).style.set_caption("Standard deviation comparison of diabetic and non-diabetic patients <br><br>")

In [33]:
# Minimum values comparison of diabetic and non-diabetic patients
df.groupby("Diabetic").min().drop(columns=["PatientID"]).style.set_caption("Minimum values comparison of diabetic and non-diabetic patients <br><br>")

In [32]:
# Maximum values comparison of diabetic and non-diabetic patients
df.groupby("Diabetic").max().drop(columns=["PatientID"]).style.set_caption("Maximum values comparison of diabetic and non-diabetic patients <br><br>")

In [4]:
# Quantile fractions comparison of diabetic and non-diabetic patients
df.groupby("Diabetic").quantile(q=[0.25, 0.5, 0.75]).drop(columns=["PatientID"]).style.set_caption("Quantile fractions comparison of diabetic and non-diabetic patients <br><br>")

# Mode of the full data frame, diabetic data frame and non-diabetic data frame

* Central tendency - To identify the central tendency of your data, you can look at the mean, median, and mode. The mean is the average value of your data, and the median is the middle value of your data when it is sorted from least to greatest. The mode is the most frequent value in your data.
* Dispersion of the data - To understand the dispersion of your data, you can look at the standard deviation or range. The standard deviation is a measure of how spread out your data is around the mean. The range is the difference between the largest and smallest values in your data.
* Outliers of the data - To identify outliers in your data, you can look at the minimum, maximum, and percentiles. Outliers are values that are significantly different from the rest of the data. For example, a value that is below the 25th percentile or above the 75th percentile may be considered an outlier.

---

In [ ]:
# Data distributions across the columns for the whole dataframe vs diabetic & non-diabetic groups

columns = df.columns
nb_rows = len(columns)

fig, axes = plt.subplots(ncols=3, nrows=nb_rows, figsize=(10, 25))

for i, col in enumerate(columns):
    sns.histplot(data=df, x=col, ax=axes[i, 0], kde=True)
    axes[i, 0].set_title(f"Full data: {col}")
    
    sns.histplot(data=df_diabetic, x=col, ax=axes[i, 1], kde=True)
    axes[i, 1].set_title(f"Diabetic: {col}")
    
    sns.histplot(data=df_non_diabetic, x=col, ax=axes[i, 2], kde=True)
    axes[i, 2].set_title(f"Non-diabetic: {col}")

plt.tight_layout()
plt.show()

#### Statistical observations - full data frame:
- comments on std - if the data is widely spread or rather clustered





#### Statistical observations - diabetic vs non-diabetic data frames:
* Diabetic group has a significantly higher SerumInsulin and PlasmaGlucose levels.

* Diabetic patients tend to have more pregnancies compared to non-diabetic patients.

* Higher glucose levels are observed in diabetics, which makes sense because diabetes is linked to high blood sugar

* Blood pressure is generally higher in diabetics compared to non-diabetics

* Diabetic patients tend to have thicker skinfolds, which may indicate higher fat distribution

* Diabetics have significantly higher insulin levels, likely due to insulin resistance

* Diabetics have a higher BMI, suggesting a link between obesity and diabetes

* Higher genetic risk is associated with diabetes

* Diabetic patients are generally older, as diabetes risk increases with age


### 3.2 Features relationships

Correlation matrix - helps to find the relationship between features (columns that predict the target value) and the target (column that solves the problem to be predicted)

Strongest Diabetic Correlation relationships : 
- Pregancies (0.41)
- Age (0.34)
- Seruminsulin (0.25)
- BMI (0.21)
- DiabetesPedigree (0.17)
- Triceps Thickness (0.15)
- PlasmaGlucose (0.13)
- DiastolicBloodPressure (0.1)


The positive correlation hints to us that all of the above meantioned features have a positive relationship to our target value (Diabetes), meaning that as they increase, the possibility of having a diabetes increases as well.


Strongest correlation is the number of pregnancies, the more pregnancies, higher the chance to get diabetes.

In [ ]:
# Bivariate analysis with a correlation matrix
df.corr()

In [ ]:
# Bivariate analysis with a heatmap correlation matrix

f, ax = plt.subplots(figsize=(10, 6))
sns.heatmap(df.corr(), annot=True, fmt=".3f", linewidths=.5, ax=ax)

Multivariate analysis

### 3.2 Depper dive into

### 3.3 Closer look at the repeated PatientID

- How many non-diabetic -> diabetic and vice versa ?

### 3.4 Outliers detection

In [ ]:
* User box plots + count tables

In [ ]:
# Data distributions across the columns for the whole dataframe vs diabetic & non-diabetic groups
columns = df.columns
nb_rows = len(columns)

fig, axes = plt.subplots(ncols=3, nrows=nb_rows, figsize=(10, 25))

for i, col in enumerate(columns):

    sns.boxplot(data=df, x=col, ax=axes[i, 0])
    axes[i, 0].set_title(f"Full data: {col}")
    
    sns.boxplot(data=df_diabetic, x=col, ax=axes[i, 1])
    axes[i, 1].set_title(f"Diabetic: {col}")
    
    sns.boxplot(data=df_non_diabetic, x=col, ax=axes[i, 2])
    axes[i, 2].set_title(f"Non-diabetic: {col}")

plt.tight_layout()
plt.show()

In [18]:
def count_outliers(df):
    outlier_counts = {}

    for col in df.select_dtypes(include=['number']).columns:
        Q1 = df[col].quantile(0.25)
        Q3 = df[col].quantile(0.75)
        IQR = Q3 - Q1
        lower_bound = Q1 - 1.5 * IQR
        upper_bound = Q3 + 1.5 * IQR

        outlier_counts[col] = df[(df[col] < lower_bound) | (df[col] > upper_bound)].shape[0]

    return outlier_counts

def outliers_per_column(df):
    outliers_per_column = count_outliers(df)
    for column, count in outliers_per_column.items():
        print(f"{column} outliers: {count}")

outliers_per_column(df)

PatientID outliers: 0
Pregnancies outliers: 0
PlasmaGlucose outliers: 0
DiastolicBloodPressure outliers: 0
TricepsThickness outliers: 31
SerumInsulin outliers: 666
BMI outliers: 0
DiabetesPedigree outliers: 475
Age outliers: 1019
Diabetic outliers: 0


In [16]:
def count_outliers_zscore_per_column(df, threshold=2):
    outlier_counts = {}

    for col in df.select_dtypes(include=['number']).columns:
        z_scores = (df[col] - df[col].mean()) / df[col].std()
        outlier_counts[col] = (np.abs(z_scores) > threshold).sum()

    return outlier_counts

# Get outlier counts per column
outliers_per_column = count_outliers_zscore_per_column(df)

# Print the result
for column, count in outliers_per_column.items():
    print(f"Number of outliers in '{column}': {count}")

Number of outliers in 'PatientID': 0
Number of outliers in 'Pregnancies': 298
Number of outliers in 'PlasmaGlucose': 522
Number of outliers in 'DiastolicBloodPressure': 270
Number of outliers in 'TricepsThickness': 74
Number of outliers in 'SerumInsulin': 769
Number of outliers in 'BMI': 202
Number of outliers in 'DiabetesPedigree': 721
Number of outliers in 'Age': 1019
Number of outliers in 'Diabetic': 0


PatientID  Pregnancies  PlasmaGlucose  \
PatientID                1.000000     0.006774      -0.001858   
Pregnancies              0.006774     1.000000       0.054502   
PlasmaGlucose           -0.001858     0.054502       1.000000   
DiastolicBloodPressure   0.008746     0.043528       0.007212   
TricepsThickness        -0.002406     0.063605       0.027100   
SerumInsulin            -0.020698     0.104487       0.033545   
BMI                     -0.003156     0.086386       0.020653   
DiabetesPedigree        -0.015413     0.054240       0.009057   
Age                     -0.007096     0.136972       0.038864   
Diabetic                -0.012494     0.407315       0.128004   

                        DiastolicBloodPressure  TricepsThickness  \
PatientID                             0.008746         -0.002406   
Pregnancies                           0.043528          0.063605   
PlasmaGlucose                         0.007212          0.027100   
DiastolicBloodPressure                1.000000          0.011106   
TricepsThickness                      0.011106          1.000000   
SerumInsulin                          0.022649          0.029688   
BMI                                   0.015873          0.024745   
DiabetesPedigree                      0.014099         -0.000951   
Age                                   0.041333          0.061383   
Diabetic                              0.091307          0.152505   

                        SerumInsulin       BMI  DiabetesPedigree       Age  \
PatientID                  -0.020698 -0.003156         -0.015413 -0.007096   
Pregnancies                 0.104487  0.086386          0.054240  0.136972   
PlasmaGlucose               0.033545  0.020653          0.009057  0.038864   
DiastolicBloodPressure      0.022649  0.015873          0.014099  0.041333   
TricepsThickness            0.029688  0.024745         -0.000951  0.061383   
SerumInsulin                1.000000  0.051223          0.046324  0.088007   
BMI                         0.051223  1.000000          0.028868  0.062910   
DiabetesPedigree            0.046324  0.028868          1.000000  0.055633   
Age                         0.088007  0.062910          0.055633  1.000000   
Diabetic                    0.247375  0.210508          0.170302  0.342605   

                        Diabetic  
PatientID              -0.012494  
Pregnancies             0.407315  
PlasmaGlucose           0.128004  
DiastolicBloodPressure  0.091307  
TricepsThickness        0.152505  
SerumInsulin            0.247375  
BMI                     0.210508  
DiabetesPedigree        0.170302  
Age                     0.342605  
Diabetic                1.000000

# 4. Feature Engineering

Goals : 
* Feature selection
* Data normalization
* Outliers treatment
- cap the outliers to min and max values, but be smart about it like titanic example
- winsorization ?
* Data encoding 
- False positive vs false negative? - rather positive, because of the context, but needs to be more argumented




Some of the possible solutions include synthesizing (duplicating) the minority class using SMOTE technique / using class weighting in algorithms (class_weight) in logistic regression, random forest, XGBoost etc. In medical field such as diabetic prediction, false negatives are more harmful that false positives. Using class weights may lead to have slightly more false positives, but that's okay in this context.

# 5. MODELS EVALUATION